ccstools/man/man8/ccs-notifyd.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH CCS-NOTIFYD "8" "May 2009" "ccs-notifyd 1.6.8" "System Administration Utilities"
.SH NAME
ccs-notifyd \- Notify TOMOYO Linux's policy violation
.SH SYNOPSIS
.B ccs-notifyd
\fItimeout action\fR
.SH DESCRIPTION
This program detects policy violation in enforcing mode and notify users (e.g. send a mail). You can know the first occurence of policy violation in enforcing mode as soon as possible. You may use this program for holding your judgement of the request that violates policy in enforcing mode provided that you can login to the system and start "ccs\-queryd" command for responding within timeout seconds.
.SH EXAMPLES

# ccs\-notifyd 180 'mail admin@example.com'
.IP
Wait for 180 seconds before rejecting the request. The occurrence is notified by sending mail to admin@example.com (if SMTP service is available).
.PP
# ccs\-notifyd 0 'curl \fB\-\-data\-binary\fR @\- https://your.server/path_to_cgi'
.IP
Reject the request immediately. The occurrence is notified by executing curl command.
.SH NOTES

 You need to register either path to this program ( /usr/lib/ccs/ccs-notifyd ) or a domain for this program in /proc/ccs/manager before invoking this program if you set non-zero value for timeout.
.SH AUTHORS

 penguin-kernel _at_ I-love.SAKURA.ne.jp
.SH COPYRIGHT
Copyright \(co 2005-2009 NTT DATA CORPORATION.
.PP
This program is free software; you may redistribute it under the terms of
the GNU General Public License. This program has absolutely no warranty.
.SH "SEE ALSO"

 ccs-queryd (8)


ccstools/ChangeLog
Version 1.0   2005/11/11   First release.

Version 1.0.1 2005/12/08   Minor update release.

  makesyaoranconf.exe:
    Use versionsort() for sorting entries.

  poled.exe:
    Support "search" command.

  poled_old.exe:
    Support "search" command.
    Fix "initializer" checking bug.

  syspol.exe:
    Support editing without resetting cursor position.

Version 1.0.2 2006/02/14   Procedure review.

  savepolicy:
    Support saving "system policy" and "exception policy"
    in addition to "domain policy".

  The following programs were added.

    editpolicy:
      "syspol.exe" "poled.exe" "poled_old.exe" were integrated
      and renamed to "editpolicy".
      This program can edit "system policy", "exception policy"
      and "domain policy".
      Command key assignments were changed.

    checkpolicy:
      A policy validator taken from "poled_old.exe".
      This program was designed for detecting and fixing errors
      in "domain policy".

    loadpolicy:
      A policy reloader.
      This program was designed for loading policy from the disk
      after clearing current policy in the kernel.

    sortpolicy:
      A "domain policy" sorter.
      This program was designed to compress access logs
      generated by "ccs-auditd".
      You can use normal "sort" command for sorting
      "system policy" and "exception policy".

    make_exception.sh:
      A script to create "exception policy".

  The following programs were renamed.

    "remount.exe" was renamed to "remount_rootfs".
    "makesyaoranconf.exe" was renamed to "makesyaoranconf".

  The following programs were removed.

    "poled.exe" "poled_old.exe" "syspol.exe"
    "obsolete_chksymlink" "obsolete_chroot_su"
    "obsolete_lsdir" "obsolete_makelink" "obsolete_movlog"
    "bindtest" "logtest"  "pathnametest" "rofstest"
    "linuxrc_old"

  The following programs for testing TOMOYO Linux's kernel were added.
  They are in the kernel_test directory.

    "sakura_bind_test" "sakura_capability_test"
    "sakura_filesystem_test" "sakura_trace_test"
    "tomoyo_capability_test" "tomoyo_file_test" "tomoyo_info_test"
    "tomoyo_name_test" "tomoyo_port_test" "tomoyo_signal_test"

Version 1.1   2006/04/01   Functionality enhancement release.

  loadpolicy:
    Delete domain for loadpolicy anyway.

  findtemp:
    Now supports for detecting all nonexistent pathnames.

  savepolicy:
    Run twice inside savepolicy itself to include necessary ACLs
    for savepolicy itself.

  The following program for testing TOMOYO Linux's kernel was added.

    "testall.sh"

Version 1.1.1 2006/05/15   Functionality enhancement release.

  The following programs were added.

    ld-watch:
      Monitors /etc/ld.so.cache and updates exception policy.
      This program is used only when updating packages.

    ccs-queryd:
      Monitors /proc/ccs/policy/query for policy violation and
      asks the administrator whether to grant or reject the request.
      This program is used while and after updating packages.

Version 1.1.2 2006/06/02   Functionality enhancement release.

  The following programs were redesigned.

    editpolicy:
      Simplified domain policy handling and removed "save" key.
      All modifications are taken effect immediately.

    loadpolicy:
      Simplified domain policy handling.

    sortpolicy:
      Simplified domain policy handling.

    savepolicy:
      Save all policies by default.

  The following program was removed.

    editpolicy_offline

Version 1.1.3 2006/07/13   Functionality enhancement release.

  The following bugs were fixed.

    editpolicy:
      The "Commands =" line was too wide to show within 80x25 screen.

    checkpolicy:
      Renamed domain for "initializer" was wrong.

Version 1.2   2006/09/03   Functionality enhancement release.

  findtemp:
    Now displays all nonexistent pathnames.

  editpolicy_offline:
    Redesigned to use the same operation manner.
    Saves changes automatically when exiting.

Version 1.3   2006/11/11   First anniversary release.

  The following program was redesigned.

    checkpolicy:
      A policy validator.
      Reads policy from stdin and prints syntax errors with line numbers.

  The following programs were added.

    setprofile:
      Assigns profiles to domains.

    pathmatch:
      Reads pathname patterns and expands them.

    domainmatch:
      fgrep for /proc/ccs/policy/domain_policy .

    ccstree:
      pstree with profile numbers and domain names.

    patternize:
      Reads domain policy and patternize pathnames.

    proxy:
      A tiny TCP port forwarder, binding to local port explicitly
      to allow servers filter based on client's port numbers.

    mailauth:
      An example program for CERBERUS.

    timeauth:
      An example program for CERBERUS, similar to honey.

  The following programs were removed.
  If you need them, please take from version 1.2 .
    "remount_rootfs" "linuxrc"
    "dumplink" "dumpsymlink" "makelink" "makesymlink"

  The following program for testing TOMOYO Linux's kernel was added.

    "tomoyo_rewrite_test"

Version 1.3.1 2006/12/08   Minor update release.

  The following bug was fixed.

    editpolicy:
      PageUp/PageDown keys and screen drawings were not working well
      on some environments due to forcefully setting "TERM=linux".

  The following program for testing TOMOYO Linux's kernel was updated.

    "newns"

Version 1.3.2 2007/02/14   Usability enhancement release.

  Many tools were merged into single source code.
  Policy editor was redesigned.

Version 1.4   2007/04/01   x86_64 support release.

  The following bug was fixed.

    editpolicy:
      Domain flags was wrong if "keep_domain <kernel>" is given.

Version 1.4.1 2007/06/05   Minor update release.

  Single source code was divided into many source code.

  The following bug was fixed.

    checkpolicy:
      "keep_domain" syntax was not checked correctly.

Version 1.4.2 2007/07/13   Bug fix release.

  .init:

ccstools/src/ccs-notifyd.c
/*
 * ccs-notifyd.c
 *
 * TOMOYO Linux's utilities.
 *
 * Copyright (C) 2005-2009  NTT DATA CORPORATION
 *
 * Version: 1.6.8   2009/05/28
 *
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <syslog.h>
#include <time.h>

int main(int argc, char *argv[])
{
	const char *proc_policy_query = "/proc/ccs/query";
	int time_to_wait;
	const char *action_to_take;
	char buffer[32768];
	FILE *fp;
	int query_fd;
	unsetenv("SHELLOPTS"); /* Make sure popen() executes commands. */
	if (argc != 3) {
		printf("Usage: %s time-to-wait action-to-take\n\n", argv[0]);
		printf("This program is used for notifying the first "
		       "occurrence of policy violation in enforcing mode.\n"
		       "The time-to-wait parameter is grace time in second "
		       "before rejecting the request that caused policy "
		       "violation.\n"
		       "The action-to-take parameter is action you want to use "
		       "for notification. "
		       "This parameter is passed to system(), so escape "
		       "appropriately as needed.\n\n");
		printf("Examples:\n\n");
		printf("  %s 180 'mail admin@example.com'\n", argv[0]);
		printf("        Wait for 180 seconds before rejecting the "
		       "request. The occurrence is notified by sending mail to "
		       "admin@example.com (if SMTP service is available)."
		       "\n\n");
		printf("  %s 0 'curl --data-binary @- "
		       "https://your.server/path_to_cgi'\n", argv[0]);
		printf("        Reject the request immediately. The occurrence "
		       "is notified by executing curl command.\n\n");
		return 0;
	}
	time_to_wait = atoi(argv[1]);
	action_to_take = argv[2];
	umask(0);
	switch (fork()) {
	case 0:
		break;
	case -1:
		fprintf(stderr, "Can't fork()\n");
		return 1;
	default:
		return 0;
	}
	if (setsid() == EOF) {
		fprintf(stderr, "Can't setsid()\n");
		return 1;
	}
	switch (fork()) {
	case 0:
		break;
	case -1:
		fprintf(stderr, "Can't fork()\n");
		return 1;
	default:
		return 0;
	}
	if (chdir("/")) {
		fprintf(stderr, "Can't chdir()\n");
		return 1;
	}
	{ /* Get exclusive lock. */
		int fd = open("/proc/self/exe", O_RDONLY);
		if (flock(fd, LOCK_EX | LOCK_NB) == EOF)
			return 0;
	}
	query_fd = open(proc_policy_query, O_RDWR);
	if (query_fd == EOF) {
		proc_policy_query = "/sys/kernel/security/tomoyo/query";
		query_fd = open(proc_policy_query, O_RDWR);
	}
	if (query_fd == EOF) {
		fprintf(stderr, "You can't run this utility for this kernel."
			"\n");
		return 1;
	} else if (time_to_wait && write(query_fd, "", 0) != 0) {
		fprintf(stderr, "You need to register this program to %s to "
			"run this program.\n", "/proc/ccs/manager");
		return 1;
	}
	close(0);
	close(1);
	close(2);
	openlog("ccs-notifyd", 0,  LOG_USER);
	syslog(LOG_WARNING, "Started. (%d, %s)\n", time_to_wait,
	       action_to_take);
	while (1) {
		fd_set rfds;
		sleep(1);
		/* Wait for query. */
		FD_ZERO(&rfds);
		FD_SET(query_fd, &rfds);
		select(query_fd + 1, &rfds, NULL, NULL, NULL);
		if (!FD_ISSET(query_fd, &rfds))
			continue;
		/* Read query. */
		memset(buffer, 0, sizeof(buffer));
		if (read(query_fd, buffer, sizeof(buffer) - 1) <= 0)
			continue;
		break;
	}
	switch (fork()) {
		time_t stamp;
		char *cp1;
		char *cp2;
	case 0:
		close(query_fd);
		fp = popen(action_to_take, "w");
		if (!fp) {
			syslog(LOG_WARNING, "Can't execute %s\n",
			       action_to_take);
			closelog();
			_exit(1);
		}
		cp1 = strchr(buffer, '\n');
		if (cp1 && sscanf(cp1 + 1, "#timestamp=%lu", &stamp) == 1) {
			cp2 = strchr(cp1 + 1, ' ');
			if (cp2) {
				/* New format. */
				struct tm *tm = localtime(&stamp);
				*(cp1 + 1) = '\0';
				fprintf(fp, "%s#%04d-%02d-%02d "
					"%02d:%02d:%02d#%s\n", buffer,
					tm->tm_year + 1900, tm->tm_mon + 1,
					tm->tm_mday, tm->tm_hour, tm->tm_min,
					tm->tm_sec, cp2);
			} else {
				fprintf(fp, "%s\n", buffer);
			}
		} else {
			fprintf(fp, "%s\n", buffer);
		}
		pclose(fp);
		_exit(0);
	case -1:
		syslog(LOG_WARNING, "Can't execute %s\n", action_to_take);
		break;
	default:
		while (time_to_wait-- > 0) {
			sleep(1);
			write(query_fd, "\n", 1);
		}
	}
	close(query_fd);
	syslog(LOG_WARNING, "Terminated.\n");
	closelog();
	return 0;
}
